de-novo transcriptome assembly development report
=================================================

## 1: k-sweep in SOAPDenovoTrans with varying read depth

Reads were subsampled at 1, 2, 4, 8, and 16 x 100k pairs from trimmed, normalised, paired 100bp rice Illumina reads.
Assemblies were performed using default settings for SOAPDenovoTrans (soapdt) with varying values of *k* from 21-86 with step 4.
All calculations were run on a 24-core Intel i7 node with 100GB RAM and with files stored in tmpfs (ramdisk) for speed.

### Assembly time scales approximately linearly with read sample size

Notably, the full k-sweep with 100k reads took only around one minute.
```{r fig.width=10, fig.height=6, echo=FALSE}
data <- read.csv('soapkoptimisation.csv', as.is=T)

library(ggplot2)
# plot time taken for each optimisation (total)
time <- aggregate(time ~ d, data[,c('d', 'time')], sum)

ggplot(time, aes(x=d, y=time/60)) +
  geom_point() +
  xlab('number of reads (x100k)') +
  ylab('time to optimise k for soapdenovo (minutes)') +
  ggtitle('Time taken to sweep + run objective functions for k in soapdenovo with increasing read depth')
```

### Three objective functions

Three objective functions were used to analyse each assembly:
* _brm_ : bad read mappings
* _rba_ : conditional annotation
* _ut_: unexpressed transcripts

```{r fig.width=10, fig.height=6, echo=FALSE}

# plot distribution of each objective function

# BMRs
data$brms <- sqrt(((0 - data$brm) / 1)**2)
ggplot(data, aes(x=k, y=brms, colour=factor(d))) +
  geom_point() +
  ylab('scaled proportion of bad read mappings') +
  scale_color_brewer(palette='Set1', name='100k reads') +
  ggtitle('Distribution of scaled BRM objective function vs. k in soapdenovo assemblies with varying read depth')

# RBA
ggplot(data, aes(x=k, y=rba, colour=factor(d))) +
  geom_point() +
  ylab('number of conditional annotations') +
  scale_color_brewer(palette='Set1', name='100k reads') +
  ggtitle('Distribution of RBA objective function vs. k in soapdenovo assemblies with varying read depth')

ggplot(data, aes(x=k, y=log(rba), colour=factor(d))) +
  geom_point() +
  ylab('log number of conditional annotations') +
  scale_color_brewer(palette='Set1', name='100k reads') +
  ggtitle('Distribution of log RBA objective function vs. k in soapdenovo assemblies with varying read depth')

data$rbas <- sqrt(((26000 - data$rba) / 26000)**2)
ggplot(data, aes(x=k, y=rbas, colour=factor(d))) +
  geom_point() +
  ylab('scaled no. of conditional annotations') +
  scale_color_brewer(palette='Set1', name='100k reads') +
  ggtitle('Distribution of scaled RBA objective function vs. k in soapdenovo assemblies with varying read depth')

# UT
ggplot(data, aes(x=k, y=ut, colour=factor(d))) +
  geom_point() +
  ylab('number of unexpressed transcripts') +
  scale_color_brewer(palette='Set1', name='100k reads') +
  ggtitle('Distribution of UT objective function vs. k in soapdenovo assemblies with varying read depth')

ggplot(data, aes(x=k, y=log(ut), colour=factor(d))) +
  geom_point() +
  ylab('log number of unexpressed transcripts') +
  scale_color_brewer(palette='Set1', name='100k reads') +
  ggtitle('Distribution of log UT objective function vs. k in soapdenovo assemblies with varying read depth')

data$uts <- sqrt(((0 - data$ut) / max(data$ut))**2)
ggplot(data, aes(x=k, y=uts, colour=factor(d))) +
  geom_point() +
  ylab('scaled no. unexpressed transcripts') +
  scale_color_brewer(palette='Set1', name='100k reads') +
  ggtitle('Distribution of scaled UT objective function vs. k in soapdenovo assemblies with varying read depth')
```

### Dimension-reduction of the three objective functions
```{r fig.width=10, fig.height=6, echo=FALSE}
# plot DFO for all points
ggplot(data, aes(x=k, y=dr, colour=factor(d))) +
  geom_point() +
  ylab('distance from optimal') +
  scale_color_brewer(palette='Set1', name='100k reads') +
  ggtitle('Dimension-reduced multi objective optimisation of k in soapdenovo assemblies with varying read depth')

# contribution of each objective to the DFO
library(reshape2)
contribs <- melt(data=data[data$d==16, c('k', 'dr', 'uts', 'rbas', 'brms')], id=c('k', 'dr'))
contribs$value = as.numeric(contribs$value)
ggplot(data=contribs, aes(x=k, y=(value/dr)/3, fill=variable)) + 
  geom_bar(stat='identity') +
  ggtitle('Contribution of each objective function to DFO for different values of k with 1600k reads') +
  ylab('proportional contribution of objective function')
contribs <- melt(data=data[data$d==1, c('k', 'dr', 'uts', 'rbas', 'brms')], id=c('k', 'dr'))
contribs$value = as.numeric(contribs$value)
ggplot(data=contribs, aes(x=k, y=(value/dr)/3, fill=variable)) + 
  geom_bar(stat='identity') +
  ggtitle('Contribution of each objective function to DFO for different values of k with 100k reads') +
  ylab('proportional contribution of objective function')
```

